File: FuckLeetcode/639.解码方法-ii.cpp
/*
 * @lc app=leetcode.cn id=639 lang=cpp
 *
 * [639] 解码方法 II
 */

#include <iostream>
#include <string>
using namespace std;

// @lc code=start
#include <unordered_map>

class Solution {
public:

    std::unordered_map<string, int> sig = {
        {"1", 1},{"2", 1},{"3", 1},{"4", 1},{"5", 1},{"6", 1},{"7", 1},{"8", 1},{"9", 1},{"10", 1},{"11", 1},{"12", 1},{"13", 1},{"14", 1},{"15", 1},{"16", 1},{"17", 1},{"18", 1},{"19", 1},{"20", 1},{"21", 1},{"22", 1},{"23", 1},{"24", 1},{"25", 1},{"26", 1},
    };
    std::unordered_map<char, int> fir2num = {
        {'0', 0},{'1', 9},{'2', 6},{'3', 0},{'4', 0},{'5', 0},{'6', 0},{'7', 0},{'8', 0},{'9', 0}
    };
    std::unordered_map<char, int> sec2num = {
        {'0', 2},{'1', 2},{'2', 2},{'3', 2},{'4', 2},{'5', 2},{'6', 2},{'7', 1},{'8', 1},{'9', 1},
    };
    
    int s2candi(string s) {
        auto it = sig.find(s);
        if (it != sig.end()) {
            return it->second;
        }
        if (s.length()==1) {
            if (s[0] != '0')
                return 9;
            else
                return 0;
        }
        else if (s=="**") {
            return 15;
        }
        else if (s[0]=='*') {
            return sec2num[s[1]];
        }
        else if (s[1]=='*') {
            return fir2num[s[0]];
        }
        else {
            return 0;
        }
    }

    int numDecodings(string s) {
        auto n = s.length();
        long long* dp = new long long[n];
        for (auto i=0; i<n; i++) {
            dp[i] = 0;
        }
        dp[0] = s2candi(s.substr(0,1));
        if (n==1) {
            return dp[0];
        }
        dp[1] = dp[0]*s2candi(s.substr(1,1)) + s2candi(s.substr(0,2));
        for (auto i=2; i<n; i++) {
            dp[i] += dp[i-1]*s2candi(s.substr(i,1)) + dp[i-2]*s2candi(s.substr(i-1,2));
            dp[i] %= 1000000007;
        }
        return dp[n-1];
    }
};
// @lc code=end

File: FuckLeetcode/639.解码方法-ii.py
#
# @lc app=leetcode.cn id=639 lang=python3
#
# [639] 解码方法 II
#

# @lc code=start
from functools import lru_cache
class Solution:
    def numDecodings(self, s: str) -> int:
        sig = set([str(i) for i in range(1, 27)])
        fir2num = {
            '0':0,
            '1':9,
            '2':6,
            '3':0,
            '4':0,
            '5':0,
            '6':0,
            '7':0,
            '8':0,
            '9':0,
        }
        sec2num = {
            '0':2,
            '1':2,
            '2':2,
            '3':2,
            '4':2,
            '5':2,
            '6':2,
            '7':1,
            '8':1,
            '9':1,
        }

        n = len(s)
        dp = [0]*n

        @lru_cache(None)
        def s2candi(s):
            if s in sig:
                return 1
            # *
            # x* *x **
            if len(s) == 1:
                return 9 if s!='0'  else 0
            elif s == '**':
                return 15
            elif s[0] == '*':
                return sec2num[s[1]]
            elif s[1] == '*':
                return fir2num[s[0]]
            else:
                return 0
        dp[0] = s2candi(s[0])
        if n==1:
            return dp[0]
        # print(dp[0]*s2candi(s[1]), s2candi(s[0:2]))
        dp[1] = dp[0]*s2candi(s[1]) + s2candi(s[0:2])
        for i in range(2, n):
            dp[i] += dp[i-1]*s2candi(s[i]) + dp[i-2]*s2candi(s[i-1:i+1])
            dp[i] %= 1000000007
        # print(dp)
        return dp[-1]
# @lc code=end
